docs/JAX_for_LLM_pretraining.md

```diff
@@ -0,0 +1,486 @@
+    (loss, logits), grads = grad_fn(model, batch)
+    metrics.update(loss=loss, logits=logits, lables=batch[1])
+    optimizer.update(grads)
+```
+
++++ {"id": "5um2vkeUNckm"}
+
+Start training. It takes ~50 minutes on Colab.
+
+Note that for data parallel, we are sharding the training data along the `batch` axis using `jax.device_put` with `NamedeSharding`.
+
+We are also using the `jax.vmap` transformation to produce the target sequences faster.
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+id: Ysl6CsfENeJN
+outputId: 5dd06dca-f030-4927-a9b6-35d412da535c
+---
+model = create_model(rngs=nnx.Rngs(0))
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+metrics = nnx.MultiMetric(
+  loss=nnx.metrics.Average('loss'),
+)
+rng = jax.random.PRNGKey(0)
+
+start_prompt = "Once upon a time"
+start_tokens = tokenizer.encode(start_prompt)[:maxlen]
+generated_text = model.generate_text(
+    maxlen, start_tokens
+)
+print(f"Initial generated text:\n{generated_text}\n")
+
+
+metrics_history = {
+  'train_loss': [],
+}
+
+prep_target_batch = jax.vmap(lambda tokens: jnp.concatenate((tokens[1:], jnp.array([0]))))
+
+step = 0
+for epoch in range(num_epochs):
+    start_time = time.time()
+    for batch in text_dl:
+        if len(batch) % len(jax.devices()) != 0:
+          continue  # skip the remaining elements
+        input_batch = jnp.array(jnp.array(batch).T)
+        target_batch = prep_target_batch(input_batch)
+        train_step(model, optimizer, metrics, jax.device_put((input_batch, target_batch), NamedSharding(mesh, P('batch', None))))
+
+        if (step + 1) % 200 == 0:
+          for metric, value in metrics.compute().items():
+              metrics_history[f'train_{metric}'].append(value)
+          metrics.reset()
+
+          elapsed_time = time.time() - start_time
+          print(f"Step {step + 1}, Loss: {metrics_history['train_loss'][-1]}, Elapsed Time: {elapsed_time:.2f} seconds")
+          start_time = time.time()
+
+          generated_text = model.generate_text(
+              maxlen, start_tokens
+          )
+          print(f"Generated text:\n{generated_text}\n")
+        step += 1
+
+# Final text generation
+generated_text = model.generate_text(
+    maxlen, start_tokens
+)
+print(f"Final generated text:\n{generated_text}")
+```
+
++++ {"id": "thaLs6TD0lt5"}
+
+Visualize the training loss.
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+  height: 472
+id: B6Eg1Cz2y_iP
+outputId: 7cafe711-1ae4-4eb9-fd37-e1bde54cbfc5
+---
+import matplotlib.pyplot as plt
+plt.plot(metrics_history['train_loss'])
+plt.title('Training Loss')
+plt.xlabel('Step')
+plt.ylabel('Loss')
+plt.show()
+```
+
++++ {"id": "WB-ExEt1Zl1C"}
+
+As you can see, the model goes from generating completely random words at the beginning to generating sensible tiny stories at the end of the training. So essentially we have pretrained a small LLM to write tiny stories for us.
+
++++ {"id": "soPqiR1JNmjf"}
+
+# Saving
```

Comment:
Here and below, we shouldn't use top-level headings except at the top level. Sorry for the late review here: I noticed this when looking at the [rendered table of contents](https://jax-ai-stack--81.org.readthedocs.build/en/81/tutorials.html) once all the build errors were fixed.